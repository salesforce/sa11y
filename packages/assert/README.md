# `assert`

Provides assertAccessible API to check DOM for accessibility issues

## Usage

// TODO(doc): Generate usage docs from code (using JSDOC, TSDOC etc)

```
import { assertAccessible } from '@sa11y/assert';

// Setup DOM in the state to be tested for accessibility
...
// Assert that the current dom has no a11y issues
// Defaults to using
//  - default document context e.g. JSDOM in Jest
//  - extended ruleset from @sa11y/preset-rules and
//  - a11yResultsFormatter from @sa11y/format
await assertAccessible();

// Can be overridden to use custom dom, ruleset or formatter
//  - Specifying null will result in using JSON stringify
// Customize rules specific to jsdom
const jsdomRules = extended;
jsdomRules.rules = {
    'color-contrast': { enabled: false }, // Disable color-contrast for jsdom
};
await assertAccessible(document, jsdomRules, null);
```
