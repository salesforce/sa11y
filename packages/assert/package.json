{
    "name": "@sa11y/assert",
    "version": "0.1.3-beta.0",
    "description": "Provides assertAccessible API to check DOM for accessibility issues",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/assert#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/assert"
    },
    "keywords": [
        "accessibility",
        "automated testing",
        "axe"
    ],
    "main": "dist/assert.js",
    "types": "dist/assert.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "@sa11y/common": "0.2.0-beta.0",
        "@sa11y/format": "0.2.1-beta.0",
        "@sa11y/preset-rules": "0.2.0-beta.0",
        "axe-core": "3.5.5"
    },
    "devDependencies": {
        "@sa11y/test-utils": "0.2.0-beta.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
