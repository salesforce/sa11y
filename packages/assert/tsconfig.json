{
    "extends": "../../tsconfig.common.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist"
    },
    "include": ["src"],
    "references": [{ "path": "../preset-rules" }, { "path": "../format" }, { "path": "../test-utils" }]
}
