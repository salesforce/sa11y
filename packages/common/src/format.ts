/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from './axe';

export const errMsgHeader = 'Accessibility issues found';

/**
 * Filter to post-process a11y results from axe
 */
export interface Filter {
    (violations: AxeResults, ...args: never[]): AxeResults;
}

// TODO (refactor): constrain rule id to known rule ids e.g using string literal, keyof, in etc
//  e.g. https://stackoverflow.com/a/54061487
// const ruleIDs = getRules().map((ruleObj) => ruleObj.ruleId);
// type RuleID = keyof ruleIDs;
// type RuleID = typeof ruleIDs[number];
// Array of length 2 or greater
type MultiArray<T> = [T, T, ...T[]];

// Selectors within a frame
type BaseSelector = string;

type ShadowDomSelector = MultiArray<BaseSelector>;
type CrossTreeSelector = BaseSelector | ShadowDomSelector;
type RuleID = string;
type CssSelectors = CrossTreeSelector[];
/**
 * Exception list of map of rule to corresponding css targets that needs to be filtered from a11y results.
 */
export type ExceptionList = Record<RuleID, CssSelectors>;
