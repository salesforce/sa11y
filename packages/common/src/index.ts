/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getAxeRules, getViolations } from './axe';
export { WdioAssertFunction, WdioOptions } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
export { log, useFilesToBeExempted, useCustomRules, processFiles } from './helpers';
