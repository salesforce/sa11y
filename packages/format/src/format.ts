/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Result } from 'axe-core';

/**
 * Formatter defines the function signature to format accessibility violations found by axe
 */
export interface Formatter {
    (violations: Result[]): string;
}

/**
 * Format accessibility results from axe
 * @param violations - Result list from axe
 * */
// TODO: Add handlebars template for formatting
// TODO: Add support for different output formats console(colored), plain text, HTML, xUnit
export function a11yResultsFormatter(violations: Result[]): string {
    return violations
        .map((violation) => {
            return violation.nodes
                .map((node) => {
                    const selectors = node.target.join(', ');
                    return `${violation.help} (${violation.id}): ${selectors}\n  - More info: ${violation.helpUrl}`;
                })
                .join('\n\n');
        })
        .join('\n\n');
}
