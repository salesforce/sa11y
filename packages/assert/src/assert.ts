/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { A11yConfig, extended } from '@sa11y/preset-rules';
import { a11yResultsFormatter, Formatter } from '@sa11y/format';

// Error message prefix for runtime exceptions when running axe
// TODO (refactor): Should this be exported? Can private variables be imported in tests?
//  Search for es6 module import version of https://www.npmjs.com/package/rewire
export const axeRuntimeExceptionMsgPrefix = 'Error running accessibility checks using axe:';

/**
 * Type def for context that can be checked for accessibility.
 * Limiting to this subset from all options supported by axe for ease of use and maintenance.
 */
export type a11yCheckableContext = Document | HTMLElement;

/**
 * Checks DOM for accessibility issues and throws an error if violations are found.
 * @param context - DOM or HTMLElement to be tested for accessibility
 * @param rules - A11yConfig preset rule to use, defaults to extended
 * @param formatter - Function to format a11y violations. Passing null will format using JSON stringify.
 * @throws error - with the accessibility issues found, does not return any value
 * */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export async function assertAccessible(
    context: a11yCheckableContext = document,
    rules: A11yConfig = extended,
    formatter: Formatter = a11yResultsFormatter
) {
    let violations;
    try {
        const results = await axe.run(context as axe.ElementContext, rules as axe.RunOptions);
        violations = results.violations;
    } catch (e) {
        throw new Error(`${axeRuntimeExceptionMsgPrefix} ${e}`);
    }
    if (violations.length > 0) {
        // TODO (improvement): Add run-time type check for Formatter interface ?
        const formattedViolations: string = formatter ? formatter(violations) : JSON.stringify(violations);
        throw new Error(formattedViolations);
    }
}
