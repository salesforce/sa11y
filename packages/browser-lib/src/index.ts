/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { exceptionListFilter, appendWcag } from '@sa11y/format';
import { defaultRuleset, registerCustomRules } from '@sa11y/preset-rules';
export { base, extended, full } from '@sa11y/preset-rules';
export const namespace = 'sa11y';

/**
 * Wrapper function to check accessibility to be invoked after the sa11y minified JS file
 * is injected into the browser
 * @param scope - Scope of the analysis as {@link A11yCheckableContext}
 * @param rules - preset sa11y rules defaulting to {@link base}
 * @param exceptionList - mapping of rule to css selectors to be filtered out using {@link exceptionListFilter}
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function checkAccessibility(
    scope = document,
    rules = defaultRuleset,
    exceptionList = {},
    addWcagInfo = false
) {
    // TODO (debug): adding type annotations to args, return type results in error:
    //  "[!] Error: Unexpected token" in both rollup-plugin-typescript2 and @rollup/plugin-typescript
    //  Compiling the index.ts file with tsc and using the dist/index.js file results
    //  in error when importing the "namespace" var. This would probably be easier to fix
    //  which could then result in getting rid of the rollup typescript plugins.

    // To register custom rules
    registerCustomRules();

    const results = await axe.run(scope || document, rules);
    const filteredResults = exceptionListFilter(results.violations, exceptionList);

    if (addWcagInfo) {
        appendWcag(filteredResults);
    }
    return JSON.stringify(filteredResults);
}
