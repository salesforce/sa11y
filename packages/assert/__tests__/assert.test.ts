/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import 'global-jsdom/lib/register'; // https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#required-globals
import { assertAccessible, axeRuntimeExceptionMsgPrefix } from '../src/assert';
import { extended, getA11yConfig, recommended } from '@sa11y/preset-rules';
import { a11yResultsFormatter } from '@sa11y/format';
import { beforeEachSetup, domWithA11yIssues, domWithNoA11yIssues, shadowDomID } from '@sa11y/test-utils';

beforeEach(() => {
    beforeEachSetup();
});

/**
 * Test util to check if given error is an a11y error
 */
function checkA11yError(e: Error): void {
    expect(e).toBeDefined();
    expect(e.toString()).not.toContain(axeRuntimeExceptionMsgPrefix);
    expect(e).toMatchSnapshot();
}

/**
 * Test util to test DOM with a11y issues
 * @param formatter - a11y results formatter
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
async function testDOMWithA11yIssues(formatter = a11yResultsFormatter) {
    document.body.innerHTML = domWithA11yIssues;
    expect.assertions(3);
    await assertAccessible(document, extended, formatter).catch((e) => {
        checkA11yError(e);
    });
}

describe('assertAccessible API', () => {
    it('should trigger axe runtime exception for non existent rule', async () => {
        const errConfig = getA11yConfig(['non-existent-rule']);
        expect.assertions(2);
        await assertAccessible(document, errConfig).catch((e) => {
            expect(e).toBeDefined();
            expect(e.toString()).toContain(axeRuntimeExceptionMsgPrefix);
        });
    });

    // eslint-disable-next-line jest/expect-expect
    it.each([recommended, extended])(
        'should throw no errors for dom with no a11y issues with config %#',
        async (config) => {
            document.body.innerHTML = domWithNoA11yIssues;
            await assertAccessible(document, config); // No error thrown
        }
    );

    it.each([
        // DOM to test, expected assertions
        [domWithNoA11yIssues, 0],
        [domWithA11yIssues, 3],
    ])(
        'should use default document, ruleset, formatter when called with no args - expecting %# assertion',
        async (testDOM: string, expectedAssertions: number) => {
            expect.assertions(expectedAssertions);
            document.body.innerHTML = testDOM;
            await assertAccessible().catch((e) => checkA11yError(e));
        }
    );

    // eslint-disable-next-line jest/expect-expect
    it('should throw an error with a11y issues found for dom with a11y issues', testDOMWithA11yIssues);

    it('should not throw error with HTML element with no a11y issues', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        const elem = document.getElementById(shadowDomID);
        expect(elem).toBeDefined();
        await assertAccessible(elem); // No error thrown
    });

    it('should throw error with HTML element with a11y issues', async () => {
        expect.assertions(5);
        document.body.innerHTML = domWithA11yIssues;
        const elements = document.getElementsByTagName('body');
        expect(elements).toHaveLength(1);
        const elem = elements[0];
        expect(elem).toBeDefined();
        await assertAccessible(elem).catch((e) => checkA11yError(e));
    });

    it.each([a11yResultsFormatter, null])('should format a11y issues using specified formatter: %#', (formatter) =>
        testDOMWithA11yIssues(formatter)
    );
});
