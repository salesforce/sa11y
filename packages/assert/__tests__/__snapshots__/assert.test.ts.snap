// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertAccessible API should format a11y issues using specified formatter: 0 1`] = `
[Error: Page must have means to bypass repeated blocks (bypass): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/bypass?application=axeAPI

Documents must have <title> element to aid in navigation (document-title): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/document-title?application=axeAPI

Document must have one main landmark (landmark-one-main): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axeAPI

Links must have discernible text (link-name): a
  - More info: https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI

Page must contain a level-one heading (page-has-heading-one): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one?application=axeAPI]
`;

exports[`assertAccessible API should format a11y issues using specified formatter: 1 1`] = `[Error: [{"id":"bypass","impact":"serious","tags":["cat.keyboard","wcag2a","wcag241","section508","section508.22.o"],"description":"Ensures each page has at least one mechanism for a user to bypass navigation and jump straight to the content","help":"Page must have means to bypass repeated blocks","helpUrl":"https://dequeuniversity.com/rules/axe/3.5/bypass?application=axeAPI","nodes":[{"any":[{"id":"internal-link-present","data":null,"relatedNodes":[],"impact":"serious","message":"No valid skip link found"},{"id":"header-present","data":null,"relatedNodes":[],"impact":"serious","message":"Page does not have a heading"},{"id":"landmark","data":null,"relatedNodes":[],"impact":"serious","message":"Page does not have a landmark region"}],"all":[],"none":[],"impact":"serious","html":"<html lang=\\"en\\"><head></head><body>\\n                            \\n                             <a href=\\"#\\"></a>\\n                            \\n                           </body></html>","target":["html"],"failureSummary":"Fix any of the following:\\n  No valid skip link found\\n  Page does not have a heading\\n  Page does not have a landmark region"}]},{"id":"document-title","impact":"serious","tags":["cat.text-alternatives","wcag2a","wcag242"],"description":"Ensures each HTML document contains a non-empty <title> element","help":"Documents must have <title> element to aid in navigation","helpUrl":"https://dequeuniversity.com/rules/axe/3.5/document-title?application=axeAPI","nodes":[{"any":[{"id":"doc-has-title","data":null,"relatedNodes":[],"impact":"serious","message":"Document does not have a non-empty <title> element"}],"all":[],"none":[],"impact":"serious","html":"<html lang=\\"en\\"><head></head><body>\\n                            \\n                             <a href=\\"#\\"></a>\\n                            \\n                           </body></html>","target":["html"],"failureSummary":"Fix any of the following:\\n  Document does not have a non-empty <title> element"}]},{"id":"landmark-one-main","impact":"moderate","tags":["cat.semantics","best-practice"],"description":"Ensures the document has a main landmark","help":"Document must have one main landmark","helpUrl":"https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axeAPI","nodes":[{"any":[],"all":[{"id":"page-has-main","data":null,"relatedNodes":[],"impact":"moderate","message":"Document does not have a main landmark"}],"none":[],"impact":"moderate","html":"<html lang=\\"en\\"><head></head><body>\\n                            \\n                             <a href=\\"#\\"></a>\\n                            \\n                           </body></html>","target":["html"],"failureSummary":"Fix all of the following:\\n  Document does not have a main landmark"}]},{"id":"link-name","impact":"serious","tags":["cat.name-role-value","wcag2a","wcag412","wcag244","section508","section508.22.a"],"description":"Ensures links have discernible text","help":"Links must have discernible text","helpUrl":"https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI","nodes":[{"any":[{"id":"has-visible-text","data":null,"relatedNodes":[],"impact":"minor","message":"Element does not have text that is visible to screen readers"},{"id":"aria-label","data":null,"relatedNodes":[],"impact":"serious","message":"aria-label attribute does not exist or is empty"},{"id":"aria-labelledby","data":null,"relatedNodes":[],"impact":"serious","message":"aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"},{"id":"role-presentation","data":null,"relatedNodes":[],"impact":"minor","message":"Element's default semantics were not overridden with role=\\"presentation\\""},{"id":"role-none","data":null,"relatedNodes":[],"impact":"minor","message":"Element's default semantics were not overridden with role=\\"none\\""}],"all":[],"none":[{"id":"focusable-no-name","data":null,"relatedNodes":[],"impact":"serious","message":"Element is in tab order and does not have accessible text"}],"impact":"serious","html":"<a href=\\"#\\"></a>","target":["a"],"failureSummary":"Fix all of the following:\\n  Element is in tab order and does not have accessible text\\n\\nFix any of the following:\\n  Element does not have text that is visible to screen readers\\n  aria-label attribute does not exist or is empty\\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\n  Element's default semantics were not overridden with role=\\"presentation\\"\\n  Element's default semantics were not overridden with role=\\"none\\""}]},{"id":"page-has-heading-one","impact":"moderate","tags":["cat.semantics","best-practice"],"description":"Ensure that the page, or at least one of its frames contains a level-one heading","help":"Page must contain a level-one heading","helpUrl":"https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one?application=axeAPI","nodes":[{"any":[],"all":[{"id":"page-has-heading-one","data":null,"relatedNodes":[],"impact":"moderate","message":"Page must have a level-one heading"}],"none":[],"impact":"moderate","html":"<html lang=\\"en\\"><head></head><body>\\n                            \\n                             <a href=\\"#\\"></a>\\n                            \\n                           </body></html>","target":["html"],"failureSummary":"Fix all of the following:\\n  Page must have a level-one heading"}]}]]`;

exports[`assertAccessible API should throw an error with a11y issues found for dom with a11y issues 1`] = `
[Error: Page must have means to bypass repeated blocks (bypass): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/bypass?application=axeAPI

Documents must have <title> element to aid in navigation (document-title): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/document-title?application=axeAPI

Document must have one main landmark (landmark-one-main): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axeAPI

Links must have discernible text (link-name): a
  - More info: https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI

Page must contain a level-one heading (page-has-heading-one): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one?application=axeAPI]
`;

exports[`assertAccessible API should throw error with HTML element with a11y issues 1`] = `
[Error: Links must have discernible text (link-name): a
  - More info: https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI]
`;

exports[`assertAccessible API should use default document, ruleset, formatter when called with no args - expecting 1 assertion 1`] = `
[Error: Page must have means to bypass repeated blocks (bypass): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/bypass?application=axeAPI

Documents must have <title> element to aid in navigation (document-title): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/document-title?application=axeAPI

Document must have one main landmark (landmark-one-main): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axeAPI

Links must have discernible text (link-name): a
  - More info: https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI

Page must contain a level-one heading (page-has-heading-one): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one?application=axeAPI]
`;
